File: easy/257. Binary Tree Paths/257. Binary Tree Paths.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}


function binaryTreePaths(root: TreeNode | null): string[] {
    if (!root)
        return []

    const run = (tree: TreeNode | null, path: string[], res: string[]) => {
        if (!tree)
            return

        path.push(`${tree.val}`)

        if (!tree.left && !tree.right)
            res.push(path.join("->"))
        else {
            run(tree.left, path, res)
            run(tree.right, path, res)
        }

        path.pop()
    }

    const res: string[] = []
    run(root, [], res)
    return res
};

File: easy/257. Binary Tree Paths/257. Binary Tree Paths.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        def run(tree: Optional[TreeNode], path: List[str], res: List[str]) -> List[str]:
            if not tree:
                return

            # path = path + "->" + str(tree.val) if len(path) else str(tree.val)
            path.append(str(tree.val))

            if not tree.left and not tree.right:
                res.append("->".join(path))
            else:
                run(tree.left, path, res)
                run(tree.right, path, res)

            path.pop()

        if not root:
            return []

        res: List[str] = []
        run(root, "", res)
        return res
